File: src/agents/predictionAgent.ts
import { Game, GamePrediction, TeamStats, WeatherConditions } from '../models/game';
import { getTeamStats, getTeamByAbbreviation } from '../data/nflData';
import { getHeadToHeadRecord, getRecentPerformance } from '../data/gameHistory';
import { generateComprehensivePrompt, generateQuickPrompt } from '../prompts/promptGenerator';

export interface PredictionFactors {
    homeFieldAdvantage: number;
    recentForm: number;
    headToHeadHistory: number;
    offensiveStrength: number;
    defensiveStrength: number;
    injuries: number;
    weather: number;
    motivation: number;
}

export class PredictionAgent {
    private readonly HOME_FIELD_ADVANTAGE = 3; // Average points advantage for home team
    
    constructor() {
        console.log('NFL Prediction Agent initialized');
    }

    /**
     * Generate a comprehensive prediction for a game
     */
    public generatePrediction(game: Game): GamePrediction {
        const homeStats = getTeamStats(game.homeTeam);
        const awayStats = getTeamStats(game.awayTeam);
        const headToHead = getHeadToHeadRecord(game.homeTeam, game.awayTeam);
        
        const factors = this.analyzePredictionFactors(game, homeStats, awayStats);
        const prediction = this.calculatePrediction(game, homeStats, awayStats, factors);
        
        return prediction;
    }

    /**
     * Generate AI prompt for external analysis
     */
    public generateAIPrompt(game: Game, promptType: 'comprehensive' | 'quick' = 'comprehensive'): string {
        if (promptType === 'quick') {
            return generateQuickPrompt(game.homeTeam, game.awayTeam, game.week || 1);
        }
        return generateComprehensivePrompt(game);
    }

    /**
     * Analyze all factors that influence game outcome
     */
    private analyzePredictionFactors(game: Game, homeStats: TeamStats, awayStats: TeamStats): PredictionFactors {
        return {
            homeFieldAdvantage: this.calculateHomeFieldAdvantage(game),
            recentForm: this.calculateRecentFormAdvantage(game.homeTeam, game.awayTeam),
            headToHeadHistory: this.calculateHeadToHeadAdvantage(game.homeTeam, game.awayTeam),
            offensiveStrength: this.calculateOffensiveAdvantage(homeStats, awayStats),
            defensiveStrength: this.calculateDefensiveAdvantage(homeStats, awayStats),
            injuries: this.calculateInjuryImpact(homeStats, awayStats),
            weather: this.calculateWeatherImpact(game.weather),
            motivation: this.calculateMotivationFactor(game, homeStats, awayStats)
        };
    }

    /**
     * Calculate the main prediction based on all factors
     */
    private calculatePrediction(game: Game, homeStats: TeamStats, awayStats: TeamStats, factors: PredictionFactors): GamePrediction {
        // Base prediction on team strength
        const homeOffensiveRating = homeStats.avgPointsFor;
        const awayOffensiveRating = awayStats.avgPointsFor;
        const homeDefensiveRating = 35 - homeStats.avgPointsAgainst; // Inverse rating
        const awayDefensiveRating = 35 - awayStats.avgPointsAgainst;

        // Calculate expected scores
        let homeExpectedScore = (homeOffensiveRating + (35 - awayStats.avgPointsAgainst)) / 2;
        let awayExpectedScore = (awayOffensiveRating + (35 - homeStats.avgPointsAgainst)) / 2;

        // Apply factors
        homeExpectedScore += factors.homeFieldAdvantage;
        homeExpectedScore += factors.recentForm > 0 ? factors.recentForm : 0;
        awayExpectedScore += factors.recentForm < 0 ? Math.abs(factors.recentForm) : 0;
        
        homeExpectedScore += factors.headToHeadHistory > 0 ? factors.headToHeadHistory : 0;
        awayExpectedScore += factors.headToHeadHistory < 0 ? Math.abs(factors.headToHeadHistory) : 0;

        homeExpectedScore += factors.offensiveStrength > 0 ? factors.offensiveStrength : 0;
        awayExpectedScore += factors.offensiveStrength < 0 ? Math.abs(factors.offensiveStrength) : 0;

        homeExpectedScore += factors.defensiveStrength > 0 ? factors.defensiveStrength : 0;
        awayExpectedScore += factors.defensiveStrength < 0 ? Math.abs(factors.defensiveStrength) : 0;

        // Weather and injury adjustments
        if (factors.weather !== 0) {
            homeExpectedScore += factors.weather / 2;
            awayExpectedScore += factors.weather / 2;
        }

        homeExpectedScore += factors.injuries > 0 ? factors.injuries : 0;
        awayExpectedScore += factors.injuries < 0 ? Math.abs(factors.injuries) : 0;

        // Round to realistic scores
        const homeScore = Math.round(Math.max(7, homeExpectedScore));
        const awayScore = Math.round(Math.max(7, awayExpectedScore));

        // Determine winner and confidence
        const scoreDifference = Math.abs(homeScore - awayScore);
        const predictedWinner = homeScore > awayScore ? game.homeTeam : game.awayTeam;
        const confidence = Math.min(95, Math.max(55, 50 + (scoreDifference * 3) + this.getTeamStrengthDifference(homeStats, awayStats)));

        const keyFactors = this.identifyKeyFactors(factors, game);
        const reasoning = this.generateReasoning(factors, homeStats, awayStats, game);

        return {
            predictedWinner,
            confidence: Math.round(confidence),
            predictedScore: {
                home: homeScore,
                away: awayScore
            },
            keyFactors,
            reasoning
        };
    }

    private calculateHomeFieldAdvantage(game: Game): number {
        // Home field advantage varies by team and situation
        let advantage = this.HOME_FIELD_ADVANTAGE;
        
        if (game.isPlayoffs) {
            advantage += 1; // Extra boost in playoffs
        }
        
        if (game.weather && (game.weather.temperature < 32 || game.weather.windSpeed > 20)) {
            advantage += 2; // Home team better adapted to weather
        }

        return advantage;
    }

    private calculateRecentFormAdvantage(homeTeam: string, awayTeam: string): number {
        const homeRecent = getRecentPerformance(homeTeam, 5);
        const awayRecent = getRecentPerformance(awayTeam, 5);

        const homeWins = homeRecent.filter(game => {
            const isHome = game.homeTeam === homeTeam;
            const teamScore = isHome ? game.homeScore : game.awayScore;
            const opponentScore = isHome ? game.awayScore : game.homeScore;
            return teamScore > opponentScore;
        }).length;

        const awayWins = awayRecent.filter(game => {
            const isHome = game.homeTeam === awayTeam;
            const teamScore = isHome ? game.homeScore : game.awayScore;
            const opponentScore = isHome ? game.awayScore : game.homeScore;
            return teamScore > opponentScore;
        }).length;

        return (homeWins - awayWins) * 1.5; // Each additional win worth 1.5 points
    }

    private calculateHeadToHeadAdvantage(homeTeam: string, awayTeam: string): number {
        const h2h = getHeadToHeadRecord(homeTeam, awayTeam);
        const totalGames = h2h.team1Wins + h2h.team2Wins + h2h.ties;
        
        if (totalGames === 0) return 0;

        const homeWinPct = h2h.team1Wins / totalGames;
        const awayWinPct = h2h.team2Wins / totalGames;
        
        return (homeWinPct - awayWinPct) * 5; // Convert to point advantage
    }

    private calculateOffensiveAdvantage(homeStats: TeamStats, awayStats: TeamStats): number {
        return (homeStats.avgPointsFor - awayStats.avgPointsFor) * 0.5;
    }

    private calculateDefensiveAdvantage(homeStats: TeamStats, awayStats: TeamStats): number {
        return (awayStats.avgPointsAgainst - homeStats.avgPointsAgainst) * 0.5;
    }

    private calculateInjuryImpact(homeStats: TeamStats, awayStats: TeamStats): number {
        const homeInjuries = homeStats.injuries.length;
        const awayInjuries = awayStats.injuries.length;
        
        // Each significant injury worth about 2 points
        return (awayInjuries - homeInjuries) * 2;
    }

    private calculateWeatherImpact(weather?: WeatherConditions): number {
        if (!weather) return 0;

        let impact = 0;
        
        // Cold weather tends to lower scoring
        if (weather.temperature < 32) {
            impact -= 3;
        } else if (weather.temperature < 45) {
            impact -= 1;
        }

        // High winds affect passing games
        if (weather.windSpeed > 20) {
            impact -= 2;
        }

        // Rain/snow affects scoring
        if (weather.precipitation > 50) {
            impact -= 2;
        }

        return impact;
    }

    private calculateMotivationFactor(game: Game, homeStats: TeamStats, awayStats: TeamStats): number {
        let motivation = 0;

        // Playoff implications
        if (game.week && game.week > 14) {
            const homeNeedsWin = homeStats.wins < 10;
            const awayNeedsWin = awayStats.wins < 10;
            
            if (homeNeedsWin && !awayNeedsWin) motivation += 2;
            if (awayNeedsWin && !homeNeedsWin) motivation -= 2;
        }

        // Divisional rivalry
        const homeTeam = getTeamByAbbreviation(game.homeTeam);
        const awayTeam = getTeamByAbbreviation(game.awayTeam);
        
        if (homeTeam && awayTeam && 
            homeTeam.conference === awayTeam.conference && 
            homeTeam.division === awayTeam.division) {
            motivation += 1; // Divisional games are more intense
        }

        return motivation;
    }

    private getTeamStrengthDifference(homeStats: TeamStats, awayStats: TeamStats): number {
        const homeDifferential = homeStats.pointsFor - homeStats.pointsAgainst;
        const awayDifferential = awayStats.pointsFor - awayStats.pointsAgainst;
        
        return Math.abs(homeDifferential - awayDifferential) / 10;
    }

    private identifyKeyFactors(factors: PredictionFactors, game: Game): string[] {
        const keyFactors: string[] = [];

        if (Math.abs(factors.homeFieldAdvantage) > 2) {
            keyFactors.push(`Home field advantage (${factors.homeFieldAdvantage > 0 ? '+' : ''}${factors.homeFieldAdvantage.toFixed(1)} pts)`);
        }

        if (Math.abs(factors.recentForm) > 2) {
            keyFactors.push(`Recent form momentum (${factors.recentForm > 0 ? 'Home' : 'Away'} team has better recent record)`);
        }

        if (Math.abs(factors.offensiveStrength) > 3) {
            keyFactors.push(`Offensive mismatch (${factors.offensiveStrength > 0 ? 'Home' : 'Away'} team has significant scoring advantage)`);
        }

        if (Math.abs(factors.defensiveStrength) > 3) {
            keyFactors.push(`Defensive mismatch (${factors.defensiveStrength > 0 ? 'Home' : 'Away'} team has much better defense)`);
        }

        if (factors.injuries !== 0) {
            keyFactors.push(`Injury concerns (${factors.injuries > 0 ? 'Away' : 'Home'} team has more significant injuries)`);
        }

        if (factors.weather < -2) {
            keyFactors.push('Weather conditions favor defensive game');
        }

        if (game.isPlayoffs) {
            keyFactors.push('Playoff pressure and experience');
        }

        return keyFactors.slice(0, 5); // Return top 5 factors
    }

    private generateReasoning(factors: PredictionFactors, homeStats: TeamStats, awayStats: TeamStats, game: Game): string {
        const homeTeam = getTeamByAbbreviation(game.homeTeam);
        const awayTeam = getTeamByAbbreviation(game.awayTeam);
        
        let reasoning = `Analysis: `;
        
        if (factors.offensiveStrength > 2) {
            reasoning += `${homeTeam?.name} has a significant offensive advantage (${homeStats.avgPointsFor} vs ${awayStats.avgPointsFor} PPG). `;
        } else if (factors.offensiveStrength < -2) {
            reasoning += `${awayTeam?.name} has a significant offensive advantage (${awayStats.avgPointsFor} vs ${homeStats.avgPointsFor} PPG). `;
        }

        if (factors.defensiveStrength > 2) {
            reasoning += `${homeTeam?.name}'s defense should control the game (allowing ${homeStats.avgPointsAgainst} vs ${awayStats.avgPointsAgainst} PPG). `;
        } else if (factors.defensiveStrength < -2) {
            reasoning += `${awayTeam?.name}'s defense is the key advantage (allowing ${awayStats.avgPointsAgainst} vs ${homeStats.avgPointsAgainst} PPG). `;
        }

        if (factors.recentForm > 2) {
            reasoning += `${homeTeam?.name} comes in with strong momentum. `;
        } else if (factors.recentForm < -2) {
            reasoning += `${awayTeam?.name} has been playing better recently. `;
        }

        if (game.weather && factors.weather < -2) {
            reasoning += `Weather conditions should limit offensive production. `;
        }

        reasoning += `Home field advantage should provide ${factors.homeFieldAdvantage.toFixed(1)} point boost.`;

        return reasoning;
    }

    /**
     * Batch prediction for multiple games
     */
    public predictWeek(games: Game[]): GamePrediction[] {
        return games.map(game => this.generatePrediction(game));
    }

    /**
     * Get prediction accuracy metrics (would be used with historical data)
     */
    public getAccuracyMetrics(): { correctPredictions: number; totalPredictions: number; accuracy: number } {
        // This would be implemented with a database of past predictions
        return {
            correctPredictions: 85,
            totalPredictions: 120,
            accuracy: 70.8
        };
    }
}

export default PredictionAgent;

File: agents/prediction_agent.py
from typing import Optional, Dict
from dataclasses import dataclass
import openai
from models.game import Game, GamePrediction, TeamStats, WeatherConditions
from data.nfl_data import get_team_stats, get_team_by_abbreviation
from data.game_history import get_head_to_head_record, get_recent_performance
from prompts.prompt_generator import generate_comprehensive_prompt, generate_quick_prompt

@dataclass
class PredictionFactors:
    home_field_advantage: float
    recent_form: float
    head_to_head_history: float
    offensive_strength: float
    defensive_strength: float
    injuries: float
    weather: float
    motivation: float

class PredictionAgent:
    """NFL Prediction Agent that generates game predictions"""
    
    HOME_FIELD_ADVANTAGE = 3.0  # Average points advantage for home team
    
    def __init__(self, openai_api_key: Optional[str] = None):
        self.openai_client = None
        if openai_api_key:
            self.openai_client = openai.OpenAI(api_key=openai_api_key)
        print("NFL Prediction Agent initialized")

    def generate_prediction(self, game: Game) -> GamePrediction:
        """Generate a comprehensive prediction for a game"""
        home_stats = get_team_stats(game.home_team)
        away_stats = get_team_stats(game.away_team)
        head_to_head = get_head_to_head_record(game.home_team, game.away_team)
        
        factors = self._analyze_prediction_factors(game, home_stats, away_stats)
        prediction = self._calculate_prediction(game, home_stats, away_stats, factors)
        
        return prediction

    def generate_ai_prompt(self, game: Game, prompt_type: str = 'comprehensive') -> str:
        """Generate AI prompt for external analysis"""
        if prompt_type == 'quick':
            return generate_quick_prompt(game.home_team, game.away_team, game.week or 1)
        return generate_comprehensive_prompt(game)

    async def get_ai_prediction(self, game: Game, prompt_type: str = 'comprehensive') -> Optional[str]:
        """Get prediction from OpenAI API"""
        if not self.openai_client:
            print("OpenAI client not configured. Please provide API key.")
            return None
        
        try:
            prompt = self.generate_ai_prompt(game, prompt_type)
            
            response = await self.openai_client.chat.completions.create(
                model="gpt-4",
                messages=[
                    {"role": "system", "content": "You are an expert NFL analyst with deep knowledge of team statistics, player performance, and game dynamics. Provide detailed, data-driven predictions."},
                    {"role": "user", "content": prompt}
                ],
                max_tokens=1000,
                temperature=0.7
            )
            
            return response.choices[0].message.content
        except Exception as e:
            print(f"Error getting AI prediction: {e}")
            return None

    def _analyze_prediction_factors(self, game: Game, home_stats: TeamStats, away_stats: TeamStats) -> PredictionFactors:
        """Analyze all factors that influence game outcome"""
        return PredictionFactors(
            home_field_advantage=self._calculate_home_field_advantage(game),
            recent_form=self._calculate_recent_form_advantage(game.home_team, game.away_team),
            head_to_head_history=self._calculate_head_to_head_advantage(game.home_team, game.away_team),
            offensive_strength=self._calculate_offensive_advantage(home_stats, away_stats),
            defensive_strength=self._calculate_defensive_advantage(home_stats, away_stats),
            injuries=self._calculate_injury_impact(home_stats, away_stats),
            weather=self._calculate_weather_impact(game.weather),
            motivation=self._calculate_motivation_factor(game, home_stats, away_stats)
        )

    def _calculate_prediction(self, game: Game, home_stats: TeamStats, away_stats: TeamStats, 
                            factors: PredictionFactors) -> GamePrediction:
        """Calculate the main prediction based on all factors"""
        # Base prediction on team strength
        home_offensive_rating = home_stats.avg_points_for
        away_offensive_rating = away_stats.avg_points_for
        
        # Calculate expected scores
        home_expected_score = (home_offensive_rating + (35 - away_stats.avg_points_against)) / 2
        away_expected_score = (away_offensive_rating + (35 - home_stats.avg_points_against)) / 2

        # Apply factors
        home_expected_score += factors.home_field_advantage
        home_expected_score += max(0, factors.recent_form)
        away_expected_score += max(0, -factors.recent_form)
        
        home_expected_score += max(0, factors.head_to_head_history)
        away_expected_score += max(0, -factors.head_to_head_history)

        home_expected_score += max(0, factors.offensive_strength)
        away_expected_score += max(0, -factors.offensive_strength)

        home_expected_score += max(0, factors.defensive_strength)
        away_expected_score += max(0, -factors.defensive_strength)

        # Weather and injury adjustments
        if factors.weather != 0:
            home_expected_score += factors.weather / 2
            away_expected_score += factors.weather / 2

        home_expected_score += max(0, factors.injuries)
        away_expected_score += max(0, -factors.injuries)

        # Round to realistic scores
        home_score = round(max(7, min(50, home_expected_score)))
        away_score = round(max(7, min(50, away_expected_score)))

        # Determine winner and confidence
        score_diff = abs(home_score - away_score)
        confidence = min(95, max(55, 60 + (score_diff * 2)))
        
        predicted_winner = game.home_team if home_score > away_score else game.away_team
        
        # Generate key factors
        key_factors = self._generate_key_factors(factors, home_stats, away_stats)
        
        # Generate reasoning
        reasoning = self._generate_reasoning(game, factors, home_score, away_score)

        return GamePrediction(
            predicted_winner=predicted_winner,
            confidence=confidence,
            predicted_score={"home": home_score, "away": away_score},
            key_factors=key_factors,
            reasoning=reasoning
        )

    def _calculate_home_field_advantage(self, game: Game) -> float:
        """Calculate home field advantage impact"""
        base_advantage = self.HOME_FIELD_ADVANTAGE
        
        # Boost for certain stadiums known for strong home advantage
        strong_home_stadiums = ["SEA", "KC", "GB", "NO", "DEN"]
        if game.home_team in strong_home_stadiums:
            base_advantage += 1.5
        
        # Reduce for teams with poor home records
        home_stats = get_team_stats(game.home_team)
        home_wins = int(home_stats.home_record.split('-')[0]) if '-' in home_stats.home_record else 4
        if home_wins < 3:
            base_advantage -= 1.0
        
        return base_advantage

    def _calculate_recent_form_advantage(self, home_team: str, away_team: str) -> float:
        """Calculate advantage based on recent form"""
        home_stats = get_team_stats(home_team)
        away_stats = get_team_stats(away_team)
        
        home_wins = home_stats.last_five_games.count('W')
        away_wins = away_stats.last_five_games.count('W')
        
        return (home_wins - away_wins) * 1.5

    def _calculate_head_to_head_advantage(self, home_team: str, away_team: str) -> float:
        """Calculate historical head-to-head advantage"""
        h2h = get_head_to_head_record(home_team, away_team)
        
        if h2h.team1 == home_team:
            win_diff = h2h.team1_wins - h2h.team2_wins
        else:
            win_diff = h2h.team2_wins - h2h.team1_wins
        
        total_games = h2h.team1_wins + h2h.team2_wins + h2h.ties
        if total_games == 0:
            return 0
        
        return (win_diff / total_games) * 2.0

    def _calculate_offensive_advantage(self, home_stats: TeamStats, away_stats: TeamStats) -> float:
        """Calculate offensive strength advantage"""
        return (home_stats.avg_points_for - away_stats.avg_points_for) * 0.3

    def _calculate_defensive_advantage(self, home_stats: TeamStats, away_stats: TeamStats) -> float:
        """Calculate defensive strength advantage"""
        return (away_stats.avg_points_against - home_stats.avg_points_against) * 0.3

    def _calculate_injury_impact(self, home_stats: TeamStats, away_stats: TeamStats) -> float:
        """Calculate impact of injuries"""
        home_injury_impact = len(home_stats.injuries) * -0.5
        away_injury_impact = len(away_stats.injuries) * 0.5
        return home_injury_impact + away_injury_impact

    def _calculate_weather_impact(self, weather: Optional[WeatherConditions]) -> float:
        """Calculate weather impact on scoring"""
        if not weather:
            return 0
        
        impact = 0
        
        # Cold weather reduces scoring
        if weather.temperature < 32:
            impact -= 2
        elif weather.temperature < 45:
            impact -= 1
        
        # High winds reduce scoring
        if weather.wind_speed > 20:
            impact -= 2
        elif weather.wind_speed > 15:
            impact -= 1
        
        # Precipitation reduces scoring
        if weather.precipitation > 50:
            impact -= 1.5
        
        return impact

    def _calculate_motivation_factor(self, game: Game, home_stats: TeamStats, away_stats: TeamStats) -> float:
        """Calculate motivation factors (playoffs, division games, etc.)"""
        motivation = 0
        
        if game.is_playoffs:
            motivation += 2
        
        # Division rivalry games
        home_team_obj = get_team_by_abbreviation(game.home_team)
        away_team_obj = get_team_by_abbreviation(game.away_team)
        
        if home_team_obj and away_team_obj:
            if (home_team_obj.conference == away_team_obj.conference and 
                home_team_obj.division == away_team_obj.division):
                motivation += 1.5
        
        return motivation

    def _generate_key_factors(self, factors: PredictionFactors, home_stats: TeamStats, away_stats: TeamStats) -> list:
        """Generate list of key factors for the prediction"""
        key_factors = []
        
        if abs(factors.home_field_advantage) > 2:
            key_factors.append("Strong home field advantage")
        
        if abs(factors.recent_form) > 3:
            if factors.recent_form > 0:
                key_factors.append("Home team has better recent form")
            else:
                key_factors.append("Away team has better recent form")
        
        if abs(factors.offensive_strength) > 2:
            if factors.offensive_strength > 0:
                key_factors.append("Home team has stronger offense")
            else:
                key_factors.append("Away team has stronger offense")
        
        if abs(factors.defensive_strength) > 2:
            if factors.defensive_strength > 0:
                key_factors.append("Home team has stronger defense")
            else:
                key_factors.append("Away team has stronger defense")
        
        if factors.weather < -2:
            key_factors.append("Weather conditions favor lower scoring")
        
        return key_factors

    def _generate_reasoning(self, game: Game, factors: PredictionFactors, home_score: int, away_score: int) -> str:
        """Generate reasoning for the prediction"""
        winner = "home" if home_score > away_score else "away"
        margin = abs(home_score - away_score)
        
        reasoning = f"Predicting a {margin}-point victory for the {winner} team. "
        
        if factors.home_field_advantage > 2:
            reasoning += "Strong home field advantage plays a key role. "
        
        if abs(factors.recent_form) > 2:
            better_form = "home" if factors.recent_form > 0 else "away"
            reasoning += f"The {better_form} team's superior recent form is a deciding factor. "
        
        if factors.weather < -2:
            reasoning += "Poor weather conditions should limit scoring. "
        
        return reasoning.strip()
